### ISSUE TITLE
Does the detached header support patch really works?

### ISSUE THREAD

**Issue by mxfm** [johnlane/grub#24]
Hi!

I have (maybe strange) but straightforward question - does the detached LUKS header patch really worked? How well was it tested?

I work on keeping up-to-date five grub crypto enhacement patches (detached header support, keyfiles, multiple password attempts, plain dm-crypt support and UUID hyphens support) from this repo in AUR repository of Arch Linux distribution. I personally used only plain dm-crypt support patch, but kept in AUR repo all five of them. Recently upstream made significant changes in the code which affected all these patches, so I started to fix them. After fixing and making compile the first (detached header support) patch I realized that probably the initial version of the patch was not tested well because it does not work in my test case (or my test case is wrong). 

The test case which was used to detect whether the patch works is follows. I created dummy 1GB luks v1 partition ('cryptsetup --type luks1 luksFormat /dev/sdc6') with a filesystem (btrfs and ext4 were tried), a dummy file, and a header backup on a separate partition. Then I killed the on-disk header with the coomand 'cryptsetup luksErase /dev/sdc6', leaving backup header only. I downloaded grub 2.04 (the last stable version) and these crypto enhacement patches as of March 2018 (with a small and insignificant modification to compile them).

When I boot in that version of grub (patched and installed on boot partition) the 'cryptomount' command is able to load detached header, open keyslot and get password. What does not work is that the encrypted partition does not mount - the grub command 'ls (crypto0)' shows that no file system is detected (I tested btrfs and ext4 in different cases). When I try to mount that partition directly (with ondisk header) - cryptomount correctly reports that 'no keyslots contain such password' (which is true because the single keyslot was destroyed). When I boot back to OS I am able to open that luks v1 partition and mount it. So, on the one hand the original patch seems to work (it can load header, get correct password and create pseudodevice 'crypto0'), but on the other hand it really does not - because there are no fs visible on that device.

Which use cases does the patch really consider? Does it treat detached header the same as backup header (perhaps I missed something and detached/backup headers are different things)? Does someone used this patch in practice? What was your use case?

**Comment by johnlane** [comment]
Hi, the LUKS functionality does work. I use it daily on at least six machines ranging from servers to laptops. I've used it with ext4, lvm and, most recently, zfs. I also use Arch.

I've used whole disks that are LUKSFormatted with the detached header from the start, specifying a sector offset of zero so that there is no unused header space at the front (so encrypted the data starts at sector 0). Like this
```
cryptsetup luksFormat --header my_header_file --align-payload 0 /dev/sdX
```

This is a total guess but the way you're doing it the data isn't at sector 0 and that may be causing you a problem.

The real use case for this was for LUKS volumes formatted with a detached header rather than copying a header from an existing volume before erasing it (in which case the "backup" might still be there, defeating the purpose. I don't know). It wasn't intended to be used that way.

There is an offset command-line option but as far as I can recall, and I can't check right now, that only applied to plain mode.

You _could_ extract the volume key from the detached header and use that with the appropriate offset and cipher details to unlock the volume using plain mode. Example 2 on the webpage does this but needs the extracted master key (I think LUKS has a command that will export this key for you).
